osf/migrations/0155_pagecounter_data.py

```diff
@@ -0,0 +1,60 @@
+# -*- coding: utf-8 -*-
+# Generated by Django 1.11.15 on 2018-11-12 17:18
+from __future__ import unicode_literals
+import progressbar
+
+from django.db import migrations
+from bulk_update.helper import bulk_update
+
+
+def reverse_func(state, schema):
+    PageCounter = state.get_model('osf', 'PageCounter')
+
+    pagecounters = PageCounter.objects.all()
+    batch = []
+    for pc in pagecounters:
+        pc.resource_id = None
+        pc.file_id = None
+        pc.version = None
+        batch.append(pc)
+
+    bulk_update(batch, update_fields=['resource_id', 'file_id', 'version'], batch_size=10000)
+
+
+def separate_pagecounter_id(state, schema):
+    """
+    Splits the data in pagecounter _id field of form action:resource_id:file_id:version into
+    four new columns: action(char), resource(fk), file(fk), version(int)
+    """
+    Guid = state.get_model('osf', 'Guid')
+    BaseFileNode = state.get_model('osf.BaseFileNode')
+    PageCounter = state.get_model('osf', 'PageCounter')
+
+    pagecounters = PageCounter.objects.all()
+    progress_bar = progressbar.ProgressBar(maxval=len(pagecounters) or 1).start()
+    batch = []
+
+    for i, pc in enumerate(pagecounters, 1):
+        id_array = pc._id.split(':')
+
+        pc.action = id_array[0]
+        pc.resource_id = Guid.objects.get(_id=id_array[1]).id
+        pc.file_id = BaseFileNode.objects.get(_id=id_array[2]).id
+        pc.version = id_array[3] if len(id_array) == 4 else None
+        batch.append(pc)
+        progress_bar.update(i)
+    bulk_update(batch, update_fields=['action', 'resource_id', 'file_id', 'version'], batch_size=10000)
+    progress_bar.finish()
```

Comment:
Was formerly in SQL but I ran it on Steve's new post-NPD staging dump and it's realllyy slow.  I modified to be more similar to Matt's pagecounter [NPD migration fix ](https://github.com/CenterForOpenScience/osf.io/pull/8841)

SQL here
```SQL
def separate_pagecounter_id(state, schema):
    """
    Splits the data in pagecounter _id field of form action:guid_id:file_id:version into
    four new columns: action(char), guid(fk), file(fk), version(int)
    """
    with connection.cursor() as cursor:
        cursor.execute(
            """
            UPDATE osf_pagecounter PC
            SET (action, guid_id, file_id, version) = (split_part(PC._id, ':', 1), G.id, F.id, NULLIF(split_part(PC._id, ':', 4), '')::int)
            FROM osf_guid G, osf_basefilenode F
            WHERE PC._id LIKE '%' || G._id || '%'
            AND PC._id LIKE '%' || F._id || '%';
            """
        )
```

Reply:
I think the only flaw with the SQL above is trying to compare to the `PC._id` in the WHERE clauses rather than the other way around, because any time you have to search on a field where the `%` is not at the end, it will have to go through the records by hand rather than using the index. 

So the following _should_ be much more performant than the previous SQL and, hopefully, the python loops. I'm testing now.

```
UPDATE osf_pagecounter PC
SET
    action = split_part(PC._id, ':', 1),
    resource_id = Guid.id,
    file_id = File.id,
    version = NULLIF(split_part(PC._id, ':', 4), '')::int
FROM osf_guid Guid, osf_basefilenode File
WHERE Guid._id = split_part(PC._id, ':', 2)
AND File._id = split_part(PC._id, ':', 3);
```